Add full UTF-8 support to RPC

laanwj opened the pull request:
(the univalue patch should be taken upstream - the RPC test and doc change should end up here. Posting it here for visibility and because this fixes an ancient issue: #2127. I think this is important to many non-English users)

This adds full UTF-8 support for both on input and output through JSON.
### bitcoin-cli usage

Before:

```
$ src/bitcoin-cli -datadir=/store/tmp/testbtc getnewaddress "рыба"
1HQCE3H87fZ1er1ExLiF5V4a1Kxf46r3J2
$ src/bitcoin-cli -datadir=/store/tmp/testbtc listaccounts
{
...  "\u00c3\u0083\u00c2\u0091\u00c3\u0082\u00c2\u0080\u00c3\u0083\u00c2\u0091\u00c3\u0082\u00c2\u008b\u00c3\u0083\u00c2\u0090\u00c3\u0082\u00c2\u00b1\u00c3\u0083\u00c2\u0090\u00c3\u0082\u00c2\u00b0": 0.00000000
}
$ src/bitcoin-cli -datadir=/store/tmp/testbtc getaccount 1HQCE3H87fZ1er1ExLiF5V4a1Kxf46r3J2
ÃÂÃÂÃÂ±ÃÂ°
```

After:

```
$ src/bitcoin-cli -regtest getaccountaddress "рыба"
mrVZ8GaURJmL3LEUu6ygU2CrUrZ8979iK2
$ src/bitcoin-cli -regtest listaccounts
{
...
  "рыба": 0.00000000
}
$ src/bitcoin-cli -regtest getaccount mrVZ8GaURJmL3LEUu6ygU2CrUrZ8979iK2                                       
рыба

```
### GUI

This also affects the GUI debug console.

Before:

![i18n_before](https://cloud.githubusercontent.com/assets/126646/14581273/2ff23f74-03e9-11e6-9be9-8504c5990e91.png)

When strings are passed directly (such as for getaccount's return argument) it works fine, but when they go through JSON formatting/parsing, it fails.

After:

![i18n_after](https://cloud.githubusercontent.com/assets/126646/14581275/35899586-03e9-11e6-8ae0-d2aede23a03c.png)
### RPC test

A test for this new functionality has been added to the `wallet.py` test.

See the commit message of the first commit for details on what exactly had to be changed in univalue.

laanwj committed:
doc: Mention full UTF-8 support in release notes

jonasschnelli commented:
Nice!
Concept ACK.

Not sure how we should treat the UniValue changes. Should we try to keep the "upstream" repository bitcoin/univalue in sync?

MarcoFalke commented:
Concept ACK. Tests look good.


> "upstream" repository bitcoin/univalue in sync

The patch goes to jgarzik/univalue as bitcoin/univalue is only for bitcoin related patches. (Still in the end we need to merge jgarzik/univalue into bitcoin/univalue because we use this as subtree.)

laanwj committed:
test: add ensure_ascii setting to AuthServiceProxy

Add a setting ensure_ascii to AuthServiceProxy. This setting,
defaulting to True (backwards compatible),
is passed through to json.dumps. If set to False, non-ASCII characters
>0x80 are not escaped. This is useful for testing server
input processing, as well as slightly more bandwidth friendly in case of
heavy unicode usage.

laanwj committed:
test: test utf-8 for labels in wallet

gmaxwell commented:
Neat.  What happens with characters like unicode direction modifiers?  Can you end up with your terminal in a weird state?

MarcoFalke commented on qa/rpc-tests/wallet.py:
```diff
@@ -1,4 +1,6 @@
 #!/usr/bin/env python2
+# coding=utf-8
+# ^^^^^^^^^^^^ TODO remove when supporting only Python3
```
Needs rebase.

laanwj commented:
> Neat. What happens with characters like unicode direction modifiers? Can you end up with your terminal in a weird state?

I don't think that's possible with just unicode?

ANSI escape sequences are a whole different story. In JSON objects they'll be escaped, but when printing strings (say, for `help`, when there is a single string result) everything is let through. Not sure whether this should be changed, but this would be a change local to `bitcoin-cli`, not the unicode framework. I'd suggest doing so in a separate pull.

Nothing new here though: It's always been possible to pass any kind of crap characters from the server (just look at my OP) to bitcoin-cli, this just makes the round-trip sane and conform to UTF-8.

dcousens commented:
concept ACK, once-over utACK (did not cross-verify UTF-8 impl.)

laanwj commented:
> Could probably use more tests for that JSONUTF8Writer...
ing 
There are few direct tests for it, but all the other tests that parses JSON with strings in it is testing it.
What would be nice is doing some fuzzing using https://github.com/laanwj/univalue/tree/2015_11_unifuzz and see that everything is accepted is valid utf-8.

gmaxwell commented:
@pstratem Any interest in fuzzing this?

sipa commented:
Needs rebase.

laanwj commented:
Rebased, removed the Python3 TODOs

laanwj committed:
Squashed 'src/univalue/' changes from 2740c4f..f32df99

f32df99 Merge branch '2016_04_unicode' into bitcoin
280b191 Merge remote-tracking branch 'jgarzik/master' into bitcoin
c9a716c Handle UTF-8
bed8dd9 Version 1.0.2.
5e7985a Merge pull request #14 from laanwj/2015_11_escape_plan

git-subtree-dir: src/univalue
git-subtree-split: f32df99e96d99ab49e5eeda16cac93747d388245

laanwj committed:
Merge commit '60ab9b200654ef0914459711cf2b22be16be3dc2'

laanwj commented:
Ok:
- Merged the unicode changes (https://github.com/jgarzik/univalue/pull/22) into https://github.com/bitcoin-core/univalue
- Replaced the direct changes to src/univalue to a subtree merge

This should be ready now.

gmaxwell commented:
ACK

laanwj commented:
It looks like the build system doesn't detect changes to univalue source files and will not automatically rebuild the library: **if you get RPC test failures concerning unicode, please build from a clean tree**

MarcoFalke commented:
```sh
make clean
make distclean
```

should take care of this, I assume.

laanwj commented:
Yes, `make clean` *should* be enough.
